### PR TITLE
Trying to improve performance for MPS

### PR DESCRIPTION
**Pull request by constantinpape**
This PR introduces two changes to try and improve the AMG performance for MPS backend:
- Implement a numpy based function for `mask_to_rle_pytorch` in `_vendored`.
- Set `points_per_batch` to `None` in the constructor and choose 16 if the backend is MPS. 

See #181 for motivation.

For me there is no difference in runtime when using the version from SAM compared to the numpy version in vendored. I tried both with my laptop (`cpu`) and a A100 GPU on the cluster (`cuda`).
Note that I tried the RLE implementation from https://www.kaggle.com/code/stainsby/fast-tested-rle/notebook but it gave wrong results. I used the RLE implementation from SO instead.

AMG Benchmark with `utils.amg.mask_to_rle_pytorch`:

| model   | device   | benchmark   |   runtime |
|:--------|:---------|:------------|----------:|
| vit_b   | cpu      | amg         |   77.9112 |
| vit_b   | cuda     | amg         |   7.02548 |

AMG Benchmark with `vendored.mask_to_rle_pytorch`:

| model   | device   | benchmark   |   runtime |
|:--------|:---------|:------------|----------:|
| vit_b   | cpu      | amg         |   77.0576 |
| vit_b   | cuda     | amg         |    7.0637 |

### CODE REVIEW

**Comment by constantinpape** [comment]
Hi @GenevieveBuckley,
I hope this fixes the performance issues with MPS now. Could you give it a try and check if it actually helps?
Thanks!

**Comment by GenevieveBuckley** [comment]
> I used the RLE implementation from SO instead

Could you link to the Stack Overflow thread? Thanks!

**Comment by GenevieveBuckley** [comment]
Summary: these changes make the amg benchmark with `vit_b` and `vit_l` roughly equal on CPU vs MPS. However, for the `vit_h` model, the amg benchmark is still a lot slower on MPS.  

Here are the benchmark timings from the `rle` branch (this PR):

### vit_h
`python benchmark.py --model_type vit_h --device cpu` / `mps`
| model   | device   | benchmark             |    runtime |
|:--------|:---------|:----------------------|-----------:|
| vit_h   | cpu      | embeddings            | 18.9906    |
| vit_h   | cpu      | prompt-p1n0           |  0.036999  |
| vit_h   | cpu      | prompt-p2n4           |  0.0369489 |
| vit_h   | cpu      | prompt-box            |  0.0359869 |
| vit_h   | cpu      | prompt-box-and-points |  0.0386569 |
| vit_h   | cpu      | amg                   | 47.025     |
|---------|----------|-----------------------|-------------|
| vit_h   | mps      | embeddings            |  13.2879    |
| vit_h   | mps      | prompt-p1n0           |   0.031723  |
| vit_h   | mps      | prompt-p2n4           |   0.02931   |
| vit_h   | mps      | prompt-box            |   0.0253892 |
| vit_h   | mps      | prompt-box-and-points |   0.028873  |
| vit_h   | mps      | amg                   | 109.321     |
| vit_h   | mps      | amg                   | 62.9395     |
| vit_h   | mps      | amg                   | 120.484     |

* I ran the amg benchmark with `vit_h` and `mps` several times, because the time taken varied a lot. 

### vit_b
`python benchmark.py --model_type vit_b --device cpu` / `mps`
| model   | device   | benchmark             |    runtime |
|:--------|:---------|:----------------------|-----------:|
| vit_b   | cpu      | embeddings            |  3.7418    |
| vit_b   | cpu      | prompt-p1n0           |  0.0377681 |
| vit_b   | cpu      | prompt-p2n4           |  0.036375  |
| vit_b   | cpu      | prompt-box            |  0.035975  |
| vit_b   | cpu      | prompt-box-and-points |  0.0358782 |
| vit_b   | cpu      | amg                   | 39.9603    |
|---------|----------|-----------------------|------------|
| vit_b   | mps      | embeddings            |  2.10778   |
| vit_b   | mps      | prompt-p1n0           |  0.032361  |
| vit_b   | mps      | prompt-p2n4           |  0.0296309 |
| vit_b   | mps      | prompt-box            |  0.0258861 |
| vit_b   | mps      | prompt-box-and-points |  0.029561  |
| vit_b   | mps      | amg                   | 41.3489    |

### vit_l
`python benchmark.py --model_type vit_l --device cpu` / `mps`
| model   | device   | benchmark             |    runtime |
|:--------|:---------|:----------------------|-----------:|
| vit_l   | cpu      | embeddings            |  9.38109   |
| vit_l   | cpu      | prompt-p1n0           |  0.0396922 |
| vit_l   | cpu      | prompt-p2n4           |  0.03947   |
| vit_l   | cpu      | prompt-box            |  0.0396121 |
| vit_l   | cpu      | prompt-box-and-points |  0.0398631 |
| vit_l   | cpu      | amg                   | 40.0773    |
|---------|----------|-----------------------|------------|
| vit_l   | mps      | embeddings            |  4.78144   |
| vit_l   | mps      | prompt-p1n0           |  0.031987  |
| vit_l   | mps      | prompt-p2n4           |  0.029345  |
| vit_l   | mps      | prompt-box            |  0.0258131 |
| vit_l   | mps      | prompt-box-and-points |  0.0297689 |
| vit_l   | mps      | amg                   | 39.9337    |

File: test/test_vendored.py [hunk 1]
```diff
@@ -3,41 +3,75 @@
...
+    def _get_mask_to_box_data(self):
+        mask_numpy = np.zeros((10, 10)).astype(bool)
         mask_numpy[7:9, 3:5] = True
-        self.mask = mask_numpy
-        self.expected_result = [3, 7, 4, 8]
+        expected_result = [3, 7, 4, 8]
+        return mask_numpy, expected_result
 
-    def test_cpu_batched_mask_to_box(self):
+    def _test_batched_mask_to_box(self, device):
```
Comment (GenevieveBuckley) [inline test/test_vendored.py:17]: I like this refactoring.

File: micro_sam/_vendored.py [hunk 2]
```diff
@@ -59,3 +62,31 @@ def batched_mask_to_box(masks: torch.Tensor) -> torch.Tensor:
         out = out[0]
 
     return out
+
+
+# segment_anything/util/amg.py
+# https://github.com/facebookresearch/segment-anything
```
Comment (GenevieveBuckley) [inline micro_sam/_vendored.py:68]: We should really say more about:
1. Why we felt we needed to put these functions here, and 
2. What condition needs to be met to remove it and go back to a more official implementation. How will someone know in future when it is possible or desirable to remove these workarounds?

And yes, I didn't write either of those things down when I added a function here 😆 Oops, I think that was a mistake and we should fix that now too.

**Comment by GenevieveBuckley** [comment]
Overall, this is an excellent result! I'm very pleased, and think this PR should be merged, ideally in time to make it into the next release.

Discussion:
It is likely that the poor performance we see on vit_h is due to the model size exceeding the amount of memory MPS has available. From the [segment-anything paper](https://arxiv.org/abs/2304.02643):
- `vit_b` (base) has 91M parameters
- `vit_l` (large) has 308M parameters
- `vit_h`(huge) has 636M parameters

I don't know if we should consider making the default model type `vit_l` (instead of `vit_h`) in micro-sam.
Pros:
- We avoid this slowdown for users with MPS
- It's a smaller model, so it's a bit faster overall for all users (regardless of whether they use cpu/cuda/mps)
Cons:
- Currently there is no fine-tuned `vit_l` model for EM or LM.
- I don't know if the performance is worse or not. (Did you or one of your colleagues assess this, as part of the work before you started work on fine-tuning models? You might already know 😄 )

Possible alternatives: 
- We could warn the user if they have both MPS and `vit_h`/`vit_h_em`/`vit_h_lm` selected.

**Comment by GenevieveBuckley** [comment]
Ok, to make a much fairer benchmark comparison, I've just re-started my computer and closed everything except python.
I have a first generation M1 macbook pro with 16GB of RAM (Ventura 13.5.1 (22G90)).

If nothing else is running simultaneously, then I don't have the same memory problem with `vit_h`. I'll open a separate issue where we can discuss whether to warn users if they might be in danger of running out of memory with MPS.

### rle branch
| model   | device   | benchmark             |    runtime |
|:--------|:---------|:----------------------|-----------:|
| vit_h   | mps      | embeddings            |  7.54463   |
| vit_h   | mps      | prompt-p1n0           |  0.035074  |
| vit_h   | mps      | prompt-p2n4           |  0.0320771 |
| vit_h   | mps      | prompt-box            |  0.0284541 |
| vit_h   | mps      | prompt-box-and-points |  0.03319   |
| vit_h   | mps      | amg                   | 33.2734    |
|---------|----------|-----------------------|------------|
| vit_h   | cpu      | embeddings            | 13.037     |
| vit_h   | cpu      | prompt-p1n0           |  0.0410733 |
| vit_h   | cpu      | prompt-p2n4           |  0.0386727 |
| vit_h   | cpu      | prompt-box            |  0.0387678 |
| vit_h   | cpu      | prompt-box-and-points |  0.040426  |
| vit_h   | cpu      | amg                   | 35.8723    |

### current dev branch

| model   | device   | benchmark             |    runtime |
|:--------|:---------|:----------------------|-----------:|
| vit_h   | mps      | embeddings            |  7.44625   |
| vit_h   | mps      | prompt-p1n0           |  0.035424  |
| vit_h   | mps      | prompt-p2n4           |  0.0320592 |
| vit_h   | mps      | prompt-box            |  0.028363  |
| vit_h   | mps      | prompt-box-and-points |  0.0332661 |
| vit_h   | mps      | amg                   | 76.1841    |
|---------|----------|-----------------------|------------|
| vit_h   | cpu      | embeddings            | 13.0751    |
| vit_h   | cpu      | prompt-p1n0           |  0.0397439 |
| vit_h   | cpu      | prompt-p2n4           |  0.0428631 |
| vit_h   | cpu      | prompt-box            |  0.0406017 |
| vit_h   | cpu      | prompt-box-and-points |  0.0418942 |
| vit_h   | cpu      | amg                   | 37.8754    |

### Summary
* Your new rle (run length encoding) implementation makes a huge difference, this is fantastic 🎉 
* The embeddings are a fair bit faster with mps (vs cpu). The 2D image examples use a reasonable amount of memory, and I haven't encountered any massive slowdowns due to exceeding the available memory. It seems clear that users should prefer to use MPS instead of cpu for this wherever possible.
* The amg (automatic mask generation) uses a LOT of memory with the mps backend, which sometimes causes huge slowdowns (performance 3.5 times slower than cpu). When you do have enough memory available, the mps backend is only about 2 seconds faster than the cpu computation. The cost/benefit here is not so clear - you have the possibility of a small speedup, but more likely you might accidentally run out of memory and cause a massive slowdown instead. I would probably prefer to use cpu instead of MPS in these cases.
* Using smaller models does not drop the amount of memory used by the amg benchmark by as much as I expected. They are all quite big memory hogs (although I only ever saw the huge slowdowns for the `vit_h` model).
* Using pytorch vs pytorch-nightly makes no difference to the benchmark timings.

**Comment by constantinpape** [comment]
Thanks for looking so thoroughly into this @GenevieveBuckley!
A few comments:
- A reason for the large memory requirement in the AMG are that many masks need to be processed (64 x 64 x 3), where each mask corresponds to a (binary) array that has the image shape. To improve this we could:
  - Profile the memory usage separately and see if there is some other function we could vendor to improve it.
  - Discard unlikely masks early.
  - (I will create an issue about this later; I would handle this independent of the current PR as it is not directly related to MPS)
- Regarding `vit_l`: we have not benchmarked it yet, to keep the experimental overhead a bit limited we only did `vit_b` and `vit_h` so far. But I want to extend the benchmarks to `vit_l` as well once we also compare to the new smaller models (see #186). For now I would leave `vit_h` as the default, and we can revisit this decision later once more data is available.
 - Using different backends for AMG versus the rest: this would complicate the design a bit. For now, I would just leave things as they are and use MPS as default as available, and see if there is an easy fix to the memory problem (see also the point above).
 
 In summary: I will go ahead and merge this :).

File: micro_sam/_vendored.py (hunk 2 above)
Comment (constantinpape) [reply micro_sam/_vendored.py:68]: I didn't see that before merging; I will go ahead and fix it up on `dev` directly ;).
